File: sdk/src/main/java/com/libreshockwave/chunks/ScoreChunk.java
package com.libreshockwave.chunks;

import com.libreshockwave.format.ChunkType;
import com.libreshockwave.io.BinaryReader;

import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

/**
 * Score chunk (VWSC).
 * Contains the timeline/score data with frames and sprite channels.
 *
 * Structure matches dirplayer-rs vm-rust/src/director/chunks/score.rs
 */
public record ScoreChunk(
    int id,
    Header header,
    List<byte[]> entries,
    ScoreFrameData frameData,
    List<FrameInterval> frameIntervals
) implements Chunk {

    @Override
    public ChunkType type() {
        return ChunkType.VWSC;
    }

    /**
     * Score chunk header (24 bytes).
     */
    public record Header(
        int totalLength,
        int unk1,
        int unk2,
        int entryCount,
        int unk3,
        int entrySizeSum
    ) {
        public static Header read(BinaryReader reader) {
            return new Header(
                reader.readI32(),
                reader.readI32(),
                reader.readI32(),
                reader.readI32(),
                reader.readI32(),
                reader.readI32()
            );
        }
    }

    /**
     * Frame data header.
     */
    public record FrameDataHeader(
        int frameCount,
        int spriteRecordSize,
        int numChannels
    ) {}

    /**
     * Per-channel sprite data (24+ bytes per channel).
     */
    public record ChannelData(
        int spriteType,
        int ink,
        int foreColor,
        int backColor,
        int castLib,
        int castMember,
        int unk1,
        int unk2,
        int posY,
        int posX,
        int height,
        int width,
        int colorFlag,
        int foreColorG,
        int backColorG,
        int foreColorB,
        int backColorB
    ) {
        public static final ChannelData EMPTY = new ChannelData(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);

        public static ChannelData read(BinaryReader reader) {
            int spriteType = reader.readU8();
            int ink = reader.readU8();
            int foreColor = reader.readU8();
            int backColor = reader.readU8();
            int castLib = reader.readU16();
            int castMember = reader.readU16();
            int unk1 = reader.readU16();
            int unk2 = reader.readU16();
            int posY = reader.readU16();
            int posX = reader.readU16();
            int height = reader.readU16();
            int width = reader.readU16();

            // Extended color data
            int unk3 = reader.readU8();
            int colorFlag = (unk3 & 0xF0) >> 4;
            reader.readU8(); // unk4
            reader.readU8(); // unk5
            reader.readU8(); // unk6
            int foreColorG = reader.readU8();
            int backColorG = reader.readU8();
            int foreColorB = reader.readU8();
            int backColorB = reader.readU8();

            return new ChannelData(
                spriteType, ink, foreColor, backColor,
                castLib, castMember, unk1, unk2,
                posY, posX, height, width,
                colorFlag, foreColorG, backColorG, foreColorB, backColorB
            );
        }

        public boolean isEmpty() {
            return spriteType == 0 && ink == 0 && foreColor == 0 && backColor == 0
                && castLib == 0 && castMember == 0 && posY == 0 && posX == 0
                && height == 0 && width == 0;
        }
    }

    /**
     * Parsed frame data from Entry[0].
     */
    public record ScoreFrameData(
        FrameDataHeader header,
        byte[] decompressedData,
        List<FrameChannelEntry> frameChannelData
    ) {
        public static final ScoreFrameData EMPTY = new ScoreFrameData(
            new FrameDataHeader(0, 24, 0),
            new byte[0],
            new ArrayList<>()
        );
    }

    /**
     * Individual frame/channel data entry.
     */
    public record FrameChannelEntry(
        int frameIndex,
        int channelIndex,
        ChannelData data
    ) {}

    /**
     * Primary interval (44 bytes) - defines frame range and channel.
     */
    public record FrameIntervalPrimary(
        int startFrame,
        int endFrame,
        int unk0,
        int unk1,
        int channelIndex,
        int unk2,
        int unk3,
        int unk4,
        int unk5,
        int unk6,
        int unk7,
        int unk8
    ) {
        public static FrameIntervalPrimary read(BinaryReader reader) {
            return new FrameIntervalPrimary(
                reader.readI32(),
                reader.readI32(),
                reader.readI32(),
                reader.readI32(),
                reader.readI32(),
                reader.readU16(),
                reader.readI32(),
                reader.readU16(),
                reader.readI32(),
                reader.readI32(),
                reader.readI32(),
                reader.readI32()
            );
        }
    }

    /**
     * Secondary interval (8 bytes) - references behavior script.
     */
    public record FrameIntervalSecondary(
        int castLib,
        int castMember,
        int unk0
    ) {
        public static FrameIntervalSecondary read(BinaryReader reader) {
            return new FrameIntervalSecondary(
                reader.readU16(),
                reader.readU16(),
                reader.readI32()
            );
        }
    }

    /**
     * Combined frame interval with primary and optional secondary.
     */
    public record FrameInterval(
        FrameIntervalPrimary primary,
        FrameIntervalSecondary secondary
    ) {}

    public int getFrameCount() {
        return frameData != null ? frameData.header().frameCount() : 0;
    }

    public int getChannelCount() {
        return frameData != null ? frameData.header().numChannels() : 0;
    }

    public static ScoreChunk read(BinaryReader reader, int id, int version) {
        reader.setOrder(ByteOrder.BIG_ENDIAN);

        if (reader.bytesLeft() < 24) {
            return createEmpty(id);
        }

        // Read header
        Header header = Header.read(reader);

        // Read offsets table
        List<Integer> offsets = new ArrayList<>();
        for (int i = 0; i <= header.entryCount(); i++) {
            if (reader.bytesLeft() < 4) break;
            offsets.add(reader.readI32());
        }

        // Validate offsets
        if (offsets.size() != header.entryCount() + 1) {
            return createEmpty(id);
        }

        for (int j = 0; j < offsets.size() - 1; j++) {
            if (offsets.get(j) > offsets.get(j + 1)) {
                return createEmpty(id);
            }
        }

        // Read all entries
        List<byte[]> entries = new ArrayList<>();
        for (int i = 0; i < header.entryCount(); i++) {
            int currentOffset = offsets.get(i);
            int nextOffset = offsets.get(i + 1);
            int length = nextOffset - currentOffset;

            if (length > 0 && reader.bytesLeft() >= length) {
                entries.add(reader.readBytes(length));
            } else {
                entries.add(new byte[0]);
            }
        }

        // Process frame data from first entry (Entry[0])
        ScoreFrameData frameData = ScoreFrameData.EMPTY;
        if (!entries.isEmpty() && entries.get(0).length > 0) {
            frameData = parseFrameData(entries.get(0));
        }

        // Process behavior attachment entries (Entry[2+])
        List<FrameInterval> frameIntervals = parseFrameIntervals(entries);

        return new ScoreChunk(id, header, entries, frameData, frameIntervals);
    }

    private static ScoreChunk createEmpty(int id) {
        return new ScoreChunk(
            id,
            new Header(0, 0, 0, 0, 0, 0),
            new ArrayList<>(),
            ScoreFrameData.EMPTY,
            new ArrayList<>()
        );
    }

    private static ScoreFrameData parseFrameData(byte[] data) {
        BinaryReader reader = new BinaryReader(data, ByteOrder.BIG_ENDIAN);

        if (reader.bytesLeft() < 20) {
            return ScoreFrameData.EMPTY;
        }

        // Read header
        int actualLength = reader.readI32();
        int unk1 = reader.readI32();
        int frameCount = reader.readI32();
        int framesVersion = reader.readU16();
        int spriteRecordSize = reader.readU16();
        int numChannels = reader.readU16();

        // Skip numChannelsDisplayed based on version
        if (framesVersion > 13) {
            reader.skip(2);
        } else {
            reader.skip(2);
        }

        FrameDataHeader header = new FrameDataHeader(frameCount, spriteRecordSize, numChannels);

        // Allocate channel data buffer with overflow protection
        long totalSizeLong = (long) frameCount * numChannels * spriteRecordSize;
        if (totalSizeLong <= 0 || totalSizeLong > 50_000_000) {
            return ScoreFrameData.EMPTY; // Sanity limit or overflow
        }
        int totalSize = (int) totalSizeLong;
        byte[] channelData = new byte[totalSize];

        // Read delta-compressed frame data
        int frameIndex = 0;
        while (!reader.eof() && frameIndex < frameCount) {
            if (reader.bytesLeft() < 2) break;

            int length = reader.readU16();
            if (length == 0) break;

            int frameLength = length - 2;
            if (frameLength > 0 && reader.bytesLeft() >= frameLength) {
                byte[] frameBytes = reader.readBytes(frameLength);
                BinaryReader frameReader = new BinaryReader(frameBytes, ByteOrder.BIG_ENDIAN);

                // Parse delta entries for this frame
                while (!frameReader.eof() && frameReader.bytesLeft() >= 4) {
                    int channelSize = frameReader.readU16();
                    int channelOffset = frameReader.readU16();

                    if (channelSize > 0 && frameReader.bytesLeft() >= channelSize) {
                        byte[] channelDelta = frameReader.readBytes(channelSize);

                        int frameOffset = frameIndex * numChannels * spriteRecordSize;
                        int destOffset = frameOffset + channelOffset;
                        int endOffset = destOffset + channelSize;

                        if (endOffset <= channelData.length) {
                            System.arraycopy(channelDelta, 0, channelData, destOffset, channelSize);
                        }
                    } else {
                        break;
                    }
                }
            }
            frameIndex++;
        }

        // Parse channel data into structured entries
        List<FrameChannelEntry> frameChannelEntries = new ArrayList<>();
        BinaryReader channelReader = new BinaryReader(channelData, ByteOrder.BIG_ENDIAN);

        for (int f = 0; f < frameCount; f++) {
            for (int c = 0; c < numChannels; c++) {
                int pos = channelReader.getPosition();

                if (channelReader.bytesLeft() >= 24) {
                    ChannelData cd = ChannelData.read(channelReader);
                    channelReader.setPosition(pos + spriteRecordSize);

                    if (!cd.isEmpty()) {
                        frameChannelEntries.add(new FrameChannelEntry(f, c, cd));
                    }
                } else {
                    break;
                }
            }
        }

        return new ScoreFrameData(header, channelData, frameChannelEntries);
    }

    private static List<FrameInterval> parseFrameIntervals(List<byte[]> entries) {
        List<FrameInterval> results = new ArrayList<>();
        int i = 2; // Start at 2, skip entries 0 and 1

        while (i < entries.size()) {
            byte[] entryBytes = entries.get(i);

            if (entryBytes.length == 0) {
                i++;
                continue;
            }

            if (entryBytes.length == 44) {
                // Primary entry (44 bytes)
                BinaryReader reader = new BinaryReader(entryBytes, ByteOrder.BIG_ENDIAN);
                FrameIntervalPrimary primary = FrameIntervalPrimary.read(reader);

                // Look ahead for secondary entries
                List<FrameIntervalSecondary> secondaries = new ArrayList<>();
                int j = i + 1;

                while (j < entries.size()) {
                    int nextSize = entries.get(j).length;

                    // Check if this could be a behavior entry (8 bytes per behavior)
                    if (nextSize >= 8 && nextSize % 8 == 0) {
                        int behaviorCount = nextSize / 8;
                        BinaryReader secReader = new BinaryReader(entries.get(j), ByteOrder.BIG_ENDIAN);

                        boolean foundValidBehavior = false;
                        for (int b = 0; b < behaviorCount; b++) {
                            int castLib = secReader.readU16();
                            int castMember = secReader.readU16();
                            int unk0 = secReader.readI32();

                            if (castLib > 0 && castMember > 0) {
                                secondaries.add(new FrameIntervalSecondary(castLib, castMember, unk0));
                                foundValidBehavior = true;
                            }
                        }

                        if (foundValidBehavior) {
                            j++;
                        } else {
                            break;
                        }
                    } else {
                        break;
                    }
                }

                // Create result entries
                if (secondaries.isEmpty()) {
                    results.add(new FrameInterval(primary, null));
                } else {
                    for (FrameIntervalSecondary secondary : secondaries) {
                        results.add(new FrameInterval(primary, secondary));
                    }
                }

                i = j;
                continue;
            }

            i++;
        }

        return results;
    }
}

File: cpp/src/chunks/ScoreChunk.cpp
#include "libreshockwave/chunks/ScoreChunk.hpp"

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <utility>

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::chunks {
namespace {

class ScopedByteOrder {
public:
    ScopedByteOrder(io::BinaryReader& reader, io::ByteOrder order)
        : reader_(reader), originalOrder_(reader.order()) {
        reader_.setOrder(order);
    }

    ~ScopedByteOrder() {
        reader_.setOrder(originalOrder_);
    }

private:
    io::BinaryReader& reader_;
    io::ByteOrder originalOrder_;
};

ScoreChunk::Header readHeader(io::BinaryReader& reader) {
    return ScoreChunk::Header{
        reader.readI32(),
        reader.readI32(),
        reader.readI32(),
        reader.readI32(),
        reader.readI32(),
        reader.readI32()
    };
}

ScoreChunk::FrameIntervalPrimary readPrimaryInterval(io::BinaryReader& reader) {
    return ScoreChunk::FrameIntervalPrimary{
        reader.readI32(),
        reader.readI32(),
        reader.readI32(),
        reader.readI32(),
        reader.readI32(),
        reader.readU16(),
        reader.readI32(),
        reader.readU16(),
        reader.readI32(),
        reader.readI32(),
        reader.readI32(),
        reader.readI32()
    };
}

ScoreChunk::FrameIntervalSecondary readSecondaryInterval(io::BinaryReader& reader) {
    return ScoreChunk::FrameIntervalSecondary{
        reader.readU16(),
        reader.readU16(),
        reader.readI32()
    };
}

std::uint8_t readU8At(const std::vector<std::uint8_t>& data, std::size_t offset) {
    return data[offset];
}

std::uint16_t readU16At(const std::vector<std::uint8_t>& data, std::size_t offset) {
    return static_cast<std::uint16_t>((static_cast<std::uint16_t>(data[offset]) << 8) |
                                      static_cast<std::uint16_t>(data[offset + 1]));
}

std::int16_t readI16At(const std::vector<std::uint8_t>& data, std::size_t offset) {
    return static_cast<std::int16_t>(readU16At(data, offset));
}

ScoreChunk::ChannelData readChannelDataAt(const std::vector<std::uint8_t>& data,
                                          std::size_t offset,
                                          int spriteRecordSize) {
    const int spriteType = readU8At(data, offset);
    const int inkByte = readU8At(data, offset + 1);
    const int ink = inkByte & 0x3F;
    const int trails = (inkByte >> 6) & 0x1;
    const int stretch = (inkByte >> 7) & 0x1;
    const int foreColor = readU8At(data, offset + 2);
    const int backColor = readU8At(data, offset + 3);
    const int castLib = readU16At(data, offset + 4);
    const int castMember = readU16At(data, offset + 6);
    const int unk1 = readU16At(data, offset + 8);
    const int unk2 = readU16At(data, offset + 10);
    const int posY = readI16At(data, offset + 12);
    const int posX = readI16At(data, offset + 14);
    const int height = readU16At(data, offset + 16);
    const int width = readU16At(data, offset + 18);

    int colorFlag = 0;
    int blendByte = 0;
    int thicknessFlags = 0;
    int foreColorG = 0;
    int backColorG = 0;
    int foreColorB = 0;
    int backColorB = 0;

    if (spriteRecordSize >= 24) {
        const int unk3 = readU8At(data, offset + 20);
        colorFlag = (unk3 & 0xF0) >> 4;
        blendByte = readU8At(data, offset + 21);
        thicknessFlags = readU8At(data, offset + 22);
    }

    if (spriteRecordSize >= 28) {
        foreColorG = readU8At(data, offset + 24);
        backColorG = readU8At(data, offset + 25);
        foreColorB = readU8At(data, offset + 26);
        backColorB = readU8At(data, offset + 27);
    }

    return ScoreChunk::ChannelData{spriteType, ink, trails, stretch, foreColor, backColor, castLib, castMember,
                                   unk1, unk2, posY, posX, height, width, colorFlag, blendByte,
                                   thicknessFlags, foreColorG, backColorG, foreColorB, backColorB};
}

ScoreChunk::ScoreFrameData parseFrameData(const std::vector<std::uint8_t>& data) {
    io::BinaryReader reader(data, io::ByteOrder::BigEndian);
    if (reader.bytesLeft() < 20) {
        return ScoreChunk::ScoreFrameData::empty();
    }

    (void)reader.readI32();
    (void)reader.readI32();
    const int frameCount = reader.readI32();
    const int framesVersion = reader.readU16();
    const int spriteRecordSize = reader.readU16();
    const int numChannels = reader.readU16();
    reader.skip(2);

    const ScoreChunk::FrameDataHeader header{frameCount, spriteRecordSize, numChannels, framesVersion};
    const int frameSize = numChannels * spriteRecordSize;
    const long long totalSizeLong = static_cast<long long>(frameCount) * static_cast<long long>(frameSize);
    if (frameCount <= 0 || frameSize <= 0 || totalSizeLong <= 0 || totalSizeLong > 50'000'000LL) {
        return ScoreChunk::ScoreFrameData::empty();
    }

    const auto totalSize = static_cast<std::size_t>(totalSizeLong);
    std::vector<std::uint8_t> channelData(totalSize, 0);

    int frameIndex = 0;
    while (!reader.eof() && frameIndex < frameCount) {
        if (reader.bytesLeft() < 2) {
            break;
        }

        const int length = reader.readU16();
        if (length == 0) {
            break;
        }

        if (frameIndex > 0) {
            const auto prevOffset = static_cast<std::size_t>(frameIndex - 1) * static_cast<std::size_t>(frameSize);
            const auto currOffset = static_cast<std::size_t>(frameIndex) * static_cast<std::size_t>(frameSize);
            std::copy_n(channelData.begin() + static_cast<std::ptrdiff_t>(prevOffset),
                        static_cast<std::size_t>(frameSize),
                        channelData.begin() + static_cast<std::ptrdiff_t>(currOffset));
        }

        const int frameLength = length - 2;
        if (frameLength > 0 && reader.bytesLeft() >= static_cast<std::size_t>(frameLength)) {
            io::BinaryReader frameReader(reader.readBytes(static_cast<std::size_t>(frameLength)), io::ByteOrder::BigEndian);
            while (!frameReader.eof() && frameReader.bytesLeft() >= 4) {
                const int channelSize = frameReader.readU16();
                const int channelOffset = frameReader.readU16();
                if (channelSize <= 0 || frameReader.bytesLeft() < static_cast<std::size_t>(channelSize)) {
                    break;
                }

                auto channelDelta = frameReader.readBytes(static_cast<std::size_t>(channelSize));
                const auto frameOffset = static_cast<std::size_t>(frameIndex) * static_cast<std::size_t>(frameSize);
                const auto destOffset = frameOffset + static_cast<std::size_t>(channelOffset);
                const auto endOffset = destOffset + static_cast<std::size_t>(channelSize);
                if (endOffset <= channelData.size()) {
                    std::copy(channelDelta.begin(),
                              channelDelta.end(),
                              channelData.begin() + static_cast<std::ptrdiff_t>(destOffset));
                }
            }
        }
        ++frameIndex;
    }

    const int mainChannelsSize = framesVersion <= 7 ? 48 : 0;
    const bool isD5 = mainChannelsSize > 0;
    std::vector<ScoreChunk::FrameChannelEntry> frameChannelEntries;
    std::vector<ScoreChunk::TempoChannelData> tempoChannelEntries;
    std::vector<ScoreChunk::PaletteChannelData> paletteChannelEntries;

    for (int frame = 0; frame < frameCount; ++frame) {
        const auto frameStart = static_cast<std::size_t>(frame) * static_cast<std::size_t>(frameSize);

        if (isD5) {
            if (frameStart + 22 <= channelData.size()) {
                const int tempo = channelData[frameStart + 21] & 0xFF;
                if (tempo > 0) {
                    tempoChannelEntries.push_back(ScoreChunk::TempoChannelData{frame, tempo});
                }
            }

            const int numSprites = frameSize > mainChannelsSize ? (frameSize - mainChannelsSize) / spriteRecordSize : 0;
            for (int sprite = 0; sprite < numSprites; ++sprite) {
                const auto pos = frameStart + static_cast<std::size_t>(mainChannelsSize + sprite * spriteRecordSize);
                if (pos + 24 > channelData.size()) {
                    break;
                }
                auto channel = readChannelDataAt(channelData, pos, spriteRecordSize);
                if (!channel.isEmpty()) {
                    frameChannelEntries.push_back(ScoreChunk::FrameChannelEntry{id::FrameIndex(frame), id::ChannelId(sprite + 6), channel});
                }
            }
        } else {
            for (int channelIndex = 0; channelIndex < numChannels; ++channelIndex) {
                const auto pos = frameStart + static_cast<std::size_t>(channelIndex * spriteRecordSize);
                if (pos + static_cast<std::size_t>(spriteRecordSize) > channelData.size()) {
                    break;
                }

                if (channelIndex == 4) {
                    if (pos + 7 <= channelData.size()) {
                        const int tempo = channelData[pos + 6] & 0xFF;
                        if (tempo > 0 && tempo <= 120) {
                            tempoChannelEntries.push_back(ScoreChunk::TempoChannelData{frame, tempo});
                        }
                    }
                } else if (channelIndex == 5) {
                    if (pos + 4 <= channelData.size()) {
                        const auto castLib = static_cast<std::int16_t>((channelData[pos] << 8) | channelData[pos + 1]);
                        const auto member = static_cast<std::int16_t>((channelData[pos + 2] << 8) | channelData[pos + 3]);
                        if (member != 0) {
                            paletteChannelEntries.push_back(ScoreChunk::PaletteChannelData{frame, castLib, member});
                        }
                    }
                } else if (pos + 24 <= channelData.size()) {
                    auto channel = readChannelDataAt(channelData, pos, spriteRecordSize);
                    if (!channel.isEmpty()) {
                        frameChannelEntries.push_back(ScoreChunk::FrameChannelEntry{id::FrameIndex(frame), id::ChannelId(channelIndex), channel});
                    }
                }
            }
        }
    }

    return ScoreChunk::ScoreFrameData{
        header,
        std::move(channelData),
        std::move(frameChannelEntries),
        std::move(tempoChannelEntries),
        std::move(paletteChannelEntries)
    };
}

std::vector<ScoreChunk::FrameInterval> parseFrameIntervals(const std::vector<std::vector<std::uint8_t>>& entries) {
    std::vector<ScoreChunk::FrameInterval> results;
    std::size_t index = 2;

    while (index < entries.size()) {
        const auto& entryBytes = entries[index];
        if (entryBytes.empty()) {
            ++index;
            continue;
        }

        if (entryBytes.size() >= 44 && entryBytes.size() <= 48) {
            io::BinaryReader reader(entryBytes, io::ByteOrder::BigEndian);
            auto primary = readPrimaryInterval(reader);

            std::vector<ScoreChunk::FrameIntervalSecondary> secondaries;
            std::size_t next = index + 1;
            while (next < entries.size()) {
                const auto nextSize = entries[next].size();
                if (nextSize < 8 || nextSize % 8 != 0) {
                    break;
                }

                const auto behaviorCount = nextSize / 8;
                io::BinaryReader secondaryReader(entries[next], io::ByteOrder::BigEndian);
                bool foundValidBehavior = false;
                for (std::size_t behavior = 0; behavior < behaviorCount; ++behavior) {
                    auto secondary = readSecondaryInterval(secondaryReader);
                    if (secondary.castLib > 0 && secondary.castMember > 0) {
                        secondaries.push_back(secondary);
                        foundValidBehavior = true;
                    }
                }

                if (!foundValidBehavior) {
                    break;
                }
                ++next;
            }

            if (secondaries.empty()) {
                results.push_back(ScoreChunk::FrameInterval{primary, std::nullopt});
            } else {
                for (const auto& secondary : secondaries) {
                    results.push_back(ScoreChunk::FrameInterval{primary, secondary});
                }
            }

            index = next;
            continue;
        }

        ++index;
    }

    return results;
}

ScoreChunk createEmptyScore(const DirectorFile* file, id::ChunkId id) {
    return ScoreChunk(file, id, ScoreChunk::Header{0, 0, 0, 0, 0, 0}, {}, ScoreChunk::ScoreFrameData::empty(), {});
}

} // namespace

ScoreChunk::ChannelData ScoreChunk::ChannelData::empty() {
    return ChannelData{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
}

ScoreChunk::ChannelData ScoreChunk::ChannelData::read(io::BinaryReader& reader, int spriteRecordSize) {
    int spriteType = reader.readU8();
    const int inkByte = reader.readU8();
    const int ink = inkByte & 0x3F;
    const int trails = (inkByte >> 6) & 0x1;
    const int stretch = (inkByte >> 7) & 0x1;
    const int foreColor = reader.readU8();
    const int backColor = reader.readU8();
    const int castLib = reader.readU16();
    const int castMember = reader.readU16();
    const int unk1 = reader.readU16();
    const int unk2 = reader.readU16();
    const int posY = reader.readI16();
    const int posX = reader.readI16();
    const int height = reader.readU16();
    const int width = reader.readU16();

    int colorFlag = 0;
    int blendByte = 0;
    int thicknessFlags = 0;
    int foreColorG = 0;
    int backColorG = 0;
    int foreColorB = 0;
    int backColorB = 0;

    if (spriteRecordSize >= 24) {
        const int unk3 = reader.readU8();
        colorFlag = (unk3 & 0xF0) >> 4;
        blendByte = reader.readU8();
        thicknessFlags = reader.readU8();
        (void)reader.readU8();
    }

    if (spriteRecordSize >= 28) {
        foreColorG = reader.readU8();
        backColorG = reader.readU8();
        foreColorB = reader.readU8();
        backColorB = reader.readU8();
    }

    return ChannelData{spriteType, ink, trails, stretch, foreColor, backColor, castLib, castMember, unk1, unk2,
                       posY, posX, height, width, colorFlag, blendByte, thicknessFlags,
                       foreColorG, backColorG, foreColorB, backColorB};
}

bool ScoreChunk::ChannelData::isEmpty() const {
    return spriteType == 0 && ink == 0 && foreColor == 0 && backColor == 0 &&
           castLib == 0 && castMember == 0 && posY == 0 && posX == 0 &&
           height == 0 && width == 0;
}

int ScoreChunk::ChannelData::resolvedForeColor() const {
    if ((colorFlag & 0x1) != 0) {
        return (foreColor << 16) | (foreColorG << 8) | foreColorB;
    }
    return foreColor;
}

int ScoreChunk::ChannelData::resolvedBackColor() const {
    if ((colorFlag & 0x2) != 0) {
        return (backColor << 16) | (backColorG << 8) | backColorB;
    }
    return backColor;
}

bool ScoreChunk::ChannelData::isForeColorRGB() const { return (colorFlag & 0x1) != 0; }
bool ScoreChunk::ChannelData::isBackColorRGB() const { return (colorFlag & 0x2) != 0; }
bool ScoreChunk::ChannelData::isFlipH() const { return (thicknessFlags & 0x20) != 0; }
bool ScoreChunk::ChannelData::isFlipV() const { return (thicknessFlags & 0x40) != 0; }

std::optional<id::CastLibId> ScoreChunk::ChannelData::castLibId() const {
    if (castLib > 0) {
        return id::CastLibId(castLib);
    }
    return std::nullopt;
}

std::optional<id::MemberId> ScoreChunk::ChannelData::memberId() const {
    if (castMember > 0) {
        return id::MemberId(castMember);
    }
    return std::nullopt;
}

ScoreChunk::ScoreFrameData ScoreChunk::ScoreFrameData::empty() {
    return ScoreFrameData{FrameDataHeader{0, 24, 0, 0}, {}, {}, {}, {}};
}

id::FrameId ScoreChunk::FrameIntervalPrimary::startFrameId() const {
    return id::FrameId(std::max(1, startFrame));
}

id::FrameId ScoreChunk::FrameIntervalPrimary::endFrameId() const {
    return id::FrameId(std::max(1, endFrame));
}

id::ChannelId ScoreChunk::FrameIntervalPrimary::channelId() const {
    return id::ChannelId(channelIndex);
}

std::optional<id::CastLibId> ScoreChunk::FrameIntervalSecondary::castLibId() const {
    if (castLib > 0) {
        return id::CastLibId(castLib);
    }
    return std::nullopt;
}

std::optional<id::MemberId> ScoreChunk::FrameIntervalSecondary::memberId() const {
    if (castMember > 0) {
        return id::MemberId(castMember);
    }
    return std::nullopt;
}

ScoreChunk::ScoreChunk(const DirectorFile* file,
                       id::ChunkId id,
                       Header header,
                       std::vector<std::vector<std::uint8_t>> entries,
                       ScoreFrameData frameData,
                       std::vector<FrameInterval> frameIntervals)
    : file_(file),
      id_(id),
      header_(header),
      entries_(std::move(entries)),
      frameData_(std::move(frameData)),
      frameIntervals_(std::move(frameIntervals)) {}

const DirectorFile* ScoreChunk::file() const { return file_; }
format::ChunkType ScoreChunk::type() const { return format::ChunkType::VWSC; }
id::ChunkId ScoreChunk::id() const { return id_; }
const ScoreChunk::Header& ScoreChunk::header() const { return header_; }
const std::vector<std::vector<std::uint8_t>>& ScoreChunk::entries() const { return entries_; }
const ScoreChunk::ScoreFrameData& ScoreChunk::frameData() const { return frameData_; }
const std::vector<ScoreChunk::FrameInterval>& ScoreChunk::frameIntervals() const { return frameIntervals_; }
int ScoreChunk::getFrameCount() const { return frameData_.header.frameCount; }
int ScoreChunk::getChannelCount() const { return frameData_.header.numChannels; }
const std::vector<std::uint8_t>& ScoreChunk::getRawChannelData() const { return frameData_.decompressedData; }
int ScoreChunk::getSpriteRecordSize() const { return frameData_.header.spriteRecordSize; }

int ScoreChunk::getFrameTempo(int frame) const {
    int result = -1;
    for (const auto& tempo : frameData_.tempoChannelData) {
        if (tempo.frameIndex <= frame) {
            result = tempo.tempo;
        } else {
            break;
        }
    }
    return result;
}

std::optional<ScoreChunk::PaletteChannelData> ScoreChunk::getFramePalette(int frame) const {
    std::optional<PaletteChannelData> result;
    for (const auto& palette : frameData_.paletteChannelData) {
        if (palette.frameIndex <= frame) {
            result = palette;
        } else {
            break;
        }
    }
    return result;
}

ScoreChunk ScoreChunk::read(const DirectorFile* file, io::BinaryReader& reader, id::ChunkId id, int version) {
    (void)version;
    ScopedByteOrder order(reader, io::ByteOrder::BigEndian);

    if (reader.bytesLeft() < 24) {
        return createEmptyScore(file, id);
    }

    const auto header = readHeader(reader);
    if (header.entryCount < 0 || header.entryCount > 10000) {
        return createEmptyScore(file, id);
    }

    std::vector<int> offsets;
    offsets.reserve(static_cast<std::size_t>(header.entryCount) + 1U);
    for (int index = 0; index <= header.entryCount; ++index) {
        if (reader.bytesLeft() < 4) {
            break;
        }
        offsets.push_back(reader.readI32());
    }

    if (offsets.size() != static_cast<std::size_t>(header.entryCount + 1)) {
        return createEmptyScore(file, id);
    }

    for (std::size_t index = 0; index + 1 < offsets.size(); ++index) {
        if (offsets[index] > offsets[index + 1]) {
            return createEmptyScore(file, id);
        }
    }

    std::vector<std::vector<std::uint8_t>> entries;
    entries.reserve(static_cast<std::size_t>(header.entryCount));
    for (int index = 0; index < header.entryCount; ++index) {
        const int length = offsets[static_cast<std::size_t>(index + 1)] - offsets[static_cast<std::size_t>(index)];
        if (length > 0 && reader.bytesLeft() >= static_cast<std::size_t>(length)) {
            entries.push_back(reader.readBytes(static_cast<std::size_t>(length)));
        } else {
            entries.emplace_back();
        }
    }

    auto frameData = ScoreFrameData::empty();
    if (!entries.empty() && !entries[0].empty()) {
        frameData = parseFrameData(entries[0]);
    }

    auto frameIntervals = parseFrameIntervals(entries);
    return ScoreChunk(file, id, header, std::move(entries), std::move(frameData), std::move(frameIntervals));
}

} // namespace libreshockwave::chunks
